src/main/java/org/jenkinsci/test/acceptance/controller/WinstoneController.java

```diff
@@ -43,29 +51,63 @@ public void addJavaOpt(String javaOpt) {
         }
     }
 
-    protected final int httpPort;
+    protected int httpPort;
 
     @Inject
     public WinstoneController(Injector i) {
-        this(i, IOUtil.randomTcpPort());
+        this(i, 0);
     }
 
     public WinstoneController(Injector i, int httpPort) {
         super(i);
         this.httpPort = httpPort;
     }
 
+    @Override
+    public void startNow() throws IOException {
+        if (httpPort == 0) {
+            if (!supportsPortFileName()) {
+                httpPort = IOUtil.randomTcpPort();
+            }
+        }
+        super.startNow();
+        this.httpPort = readPort();
+    }
+
+    @Override
+    protected void onReady() throws IOException {
+        this.httpPort = readPort();
+    }
+
+    private int readPort() throws IOException {
+        if (this.httpPort == 0 && portFile != null) {
+            String s = FileUtils.readFileToString(portFile, StandardCharsets.UTF_8);
+            try {
+                return Integer.parseInt(s);
+            } catch (NumberFormatException e) {
+                // Should not happen assuming this is called after a successful start.
+                throw new IOException("Unable to parse port from " + s + ". Jenkins did not start.");
+            }
+        } else {
+            return this.httpPort;
```

Comment:
A bit confusing. Perhaps this method should return `void` and edit `[this.]httpPort` only when necessary.